cpp/src/fil/fil.cu

```diff
@@ -636,6 +637,89 @@ int tree2fil_sparse(std::vector<fil_node_t>& nodes, int root,
   return root;
 }
 
+struct level_entry {
+  int n_branches, n_leaves;
+};
+typedef std::pair<int, int> pair_t;
+using std::tie;
+// hist has branch and leaf count given depth
+template <typename T, typename L>
+inline void tree_depth_hist(const tl::Tree<T, L>& tree,
+                            std::unordered_map<int, level_entry>& hist) {
+  std::stack<pair_t> stack;  // {tl_id, depth}
+  stack.push({tree_root(tree), 0});
+  while (!stack.empty()) {
+    int node_id, depth;
+    tie(node_id, depth) = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      hist[depth].n_branches++;
+      stack.push({tree.LeftChild(node_id), depth + 1});
+      node_id = tree.RightChild(node_id);
+      depth++;
+    }
+    hist[depth].n_leaves++;
+  }
+}
+
+// implements https://tools.ietf.org/html/draft-eastlake-fnv-17.html
+// algorithm is public domain without export controls and no patents or rights to patent
+auto fowler_noll_vo_fingerprint64(int* begin, int* end) {
```

Comment:
@teju85 What do you think, where would the best place for such a function in cuML / RAFT be?